// @vue/language-core/out/generators/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeFeatureSettings = exports.enableAllFeatures = exports.disableAllFeatures = void 0;
function disableAllFeatures(override) {
    return {
        verification: false,
        completion: false,
        semantic: false,
        navigation: false,
        structure: false,
        format: false,
        ...override,
    };
}
exports.disableAllFeatures = disableAllFeatures;
function enableAllFeatures(override) {
    return {
        verification: true,
        completion: true,
        semantic: true,
        navigation: true,
        structure: true,
        format: true,
        ...override,
    };
}
exports.enableAllFeatures = enableAllFeatures;
function mergeFeatureSettings(base, ...others) {
    const result = { ...base };
    for (const info of others) {
        for (const key in info) {
            const value = info[key];
            if (value) {
                result[key] = value;
            }
        }
    }
    return result;
}
exports.mergeFeatureSettings = mergeFeatureSettings;
